src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,475 @@
+                ent.d_name);
+            goto next;
+        }
+
+        sol_ptr_vector_insert_sorted(vec, s, kmod_conf_cmp);
+
+next:
+        r = readdir_r(dir, &ent, &entbuf);
+    }
+}
+
+static void
+uevent_read_msg(struct kmod_data *kdata, char *msg, int len)
+{
+    char *action;
+    struct sol_str_slice modalias = SOL_STR_SLICE_EMPTY;
+    int i = 0;
+
+    while (i < len) {
+        char *curr, *del;
+        size_t value_len;
+
+        curr = msg + i;
+        del = strrchr(curr, '=');
+        value_len = del - curr;
+
+        if (del && !strncmp(curr, "MODALIAS", value_len))
+            modalias = sol_str_slice_from_str(++del);
+        else if (del && !strncmp(curr, "ACTION", value_len))
+            action = ++del;
+
+        i += strlen(msg + i) + 1;
+    }
+
+    if (modalias.len && !strcmp(action, "add")) {
+        kmod_apply_value(kdata->kmod, modalias);
+    }
+}
+
+static bool
+kmod_uevent_handler(void *data, int fd, unsigned int cond)
+{
+    int len;
+    struct kmod_data *kdata = data;
+    char buffer[512];
+
+    len = recv(kdata->uevent.fd, buffer, sizeof(buffer), MSG_WAITALL);
+    if (len == -1) {
+        SOL_ERR("Could not read netlink socket. %s", sol_util_strerrora(errno));
+        return false;
+    } else {
+        uevent_read_msg(kdata, buffer, len);
+    }
+
+    return true;
+}
+
+static int
+kmod_uevent_start(struct kmod_data *kdata)
+{
+    struct sockaddr_nl nls;
+
+    memset(&nls, 0, sizeof(struct sockaddr_nl));
+    nls.nl_family = AF_NETLINK;
+    nls.nl_pid = getpid();
+    nls.nl_groups = -1;
+
+    kdata->uevent.fd = socket(PF_NETLINK, SOCK_DGRAM | SOCK_CLOEXEC,
+        NETLINK_KOBJECT_UEVENT);
+    if (kdata->uevent.fd == -1) {
+        SOL_ERR("Could not open uevent netlink socket.");
+        return -errno;
+    }
+
+    if (bind(kdata->uevent.fd, (void *)&nls, sizeof(struct sockaddr_nl))) {
+        SOL_ERR("Could not bind to uevent socket.");
+        close(kdata->uevent.fd);
+        kdata->uevent.fd = 0;
+        return -errno;
+    }
+
+    kdata->uevent.watch = sol_fd_add(kdata->uevent.fd, SOL_FD_FLAGS_IN,
+        kmod_uevent_handler, kdata);
+    return 0;
+}
+
+static void
+kmod_coldplug_find_devices(struct kmod_data *kdata, DIR *dir)
+{
+    struct dirent ent, *entbuf = NULL;
+    int r, dfd;
+
+    dfd = dirfd(dir);
+    if (dfd < 0) {
+        SOL_DBG("Could not obtain directory file descriptor");
+        return;
+    }
+
+    r = readdir_r(dir, &ent, &entbuf);
+    while (r == 0 && entbuf) {
+        struct stat st;
+
+        if (!strcmp(ent.d_name, ".") || !strcmp(ent.d_name, "..")) {
+            goto next;
+        }
+
+        if (fstatat(dfd, ent.d_name, &st, AT_SYMLINK_NOFOLLOW) == -1) {
+            SOL_ERR("Could not stat %s", ent.d_name);
+            goto next;
+        }
+
+        if (S_ISDIR(st.st_mode)) {
+            int nfd;
+            DIR *ndir;
+
+            nfd = openat(dfd, ent.d_name, O_RDONLY | O_CLOEXEC);
```

Comment:
Maybe also `O_PATH`, to ensure `dfd` isn't used for anything else than path-related operations?